python/lsst/daf/butler/datastores/s3Datastore.py

```diff
@@ -238,6 +238,11 @@ def put(self, inMemoryDataset, ref):
                                         Filename=tmpFile.name)
                 log.debug("Wrote file to %s via a temporary directory.", location.uri)
 
+        # Register a callback to try to delete the uploaded data if
+        # the ingest fails below
+        self._transaction.registerUndo("write", self.client.delete_object,
```

Comment:
@DinoBektesevic I've added this to the S3Datastore. Whether this is in the right place or not depends on whether upload_file or put_object are themselves transactional. If they fail and the object is effectively not in S3 then this is the right place. If they can partially work then instead of registerUndo we will have to do what we do in posixDatastore (use undoWith, catch the exception, reraise).

Reply:
This will work. S3 API docs claim never to [add partial files](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html) which is what boto3 eventually calls for both object and file uploads.

In my Jira comment, I was actually concerned about the upload_file calls in the `_ExtractIngestInfo` functionality. That bit has no, but definitely should, undo transactions protections in S3Datastore. However, on move actions it would require me to delete the files in the original location (whether local disk or an different S3 Bucket). For S3 Bucket then it becomes difficult to confirm file actually got deleted because the `delete_object` calls perpetually return HTML 204 response.

Reply:
Ok. I think my line does make things better and won't trigger any problems in the ingest line because transfer=None. I can leave the transaction handling for `ingest` to another time.